### ISSUE TITLE
Loading protocol fails when plugins are not installed (Trac #164)

### ISSUE THREAD

**Issue by cfobel** [wheeler-microfluidics/microdrop#166]
Plugins currently store their protocol-level and step-level data within serialized strings contained in a a global Protocol() object. When a protocol is loaded from a file, if any serialized data exists for plugins that are not available (but were installed/enabled when the protocol was created), it causes an error.

To solve this problem, plugins should be responsible for the serialiation/deserialization of their own data. If a plugin is not available when a protocol is loaded, that plugin's data will simply be ignored (perhaps displaying a warning to the user).

I propose that any plugin that wants to load/save data to/from a protocol should expose methods:

```serialize_protocol_data```
```deserialize_protocol_data```

These methods would take a protocol object as an argument and convert data in-place between serialized strings and python objects.

Migrated from http://microfluidics.utoronto.ca/microdrop/ticket/164
```json
{
    "status": "closed", 
    "changetime": "2015-01-13T08:41:17", 
    "description": "Plugins currently store their protocol-level and step-level data within serialized strings contained in a a global Protocol() object. When a protocol is loaded from a file, if any serialized data exists for plugins that are not available (but were installed/enabled when the protocol was created), it causes an error.\n\nTo solve this problem, plugins should be responsible for the serialiation/deserialization of their own data. If a plugin is not available when a protocol is loaded, that plugin's data will simply be ignored (perhaps displaying a warning to the user).\n\nI propose that any plugin that wants to load/save data to/from a protocol should expose methods:\n\n{{{serialize_protocol_data}}}\n{{{deserialize_protocol_data}}}\n\nThese methods would take a protocol object as an argument and convert data in-place between serialized strings and python objects.", 
    "reporter": "ryan", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1421138477180309", 
    "component": "microdrop", 
    "summary": "Loading protocol fails when plugins are not installed", 
    "priority": "major", 
    "keywords": "", 
    "version": "0.1", 
    "time": "2014-12-08T20:24:43", 
    "milestone": "Microdrop 1.0", 
    "owner": "", 
    "type": "enhancement"
}
```

**Comment by cfobel** [comment]
Trac update at 20150108T21:42:07: ryan commented:

> It's not necessary for plugins to do their own serialization/deserialization to prevent this error. This is actually an independent feature that could be added if there is a need (e.g., if we want to allow for non-python based plugins).
> 
> For now, it should be sufficient to enable the loading of protocols without having all of the plugins installed/enabled. The program should provide a warning saying that any plugin data from these missing plugins will be ignored. It may also be nice to offer to delete this extra data in order to prevent warnings from being displayed in the future.

**Comment by cfobel** [comment]
Trac update at 20150113T08:41:17: 
* ryan commented:

> Fixed in commit changeset 2db2734.


* ryan changed resolution from "" to "fixed"
* ryan changed status from "new" to "closed"
